src/urllib3/request.py

```diff
@@ -84,6 +86,16 @@ def request(
 
         urlopen_kw["request_url"] = url
 
+        if json is not None and body is not None:
+            raise TypeError(
+                "request got values for both 'json' and 'body', can only specify one. Prefer json"
+            )
+
+        if json is not None:
+            if headers is None:
+                headers = self.headers.setdefault("Content-Type", "application/json")  # type: ignore
```

Comment:
We need to support `HTTPHeaderDict`, so you'll have to check if the header is defined before setting it instead of using `setdefault `.

Reply:
Hmm, So the logic should be , If `headers is None`, then `setdefault`
Else if headers is HTTPHeaderDict, then check if it has Content-Type (No, then `setdefault`)
else if it is not HTTPHeaderDict and Not None then throw error
 ?

Reply:
Ok, I just realized what `setdefault` does on empty dict.

Reply:
@pquentin  or @shazow . I'm a little confused on this part

Does ``defined before``means like this.
```python
import urllib3
header = urllib3.HTTPHeaderDict(cookie="foo, bar")
http = urllib3.PoolManager(headers=header)
data = {"attribute": "value"}
r = http.request(
    "POST",
    "http://httpbin.org/post",
    json=data
)
```
So self.headers can be `HTTPHeaderDict` or `{}`
In either case, we can by default set it as 
```python
if headers is None:
    headers  = self.headers
    headers["Content-Type"] = "application/json"
```

Or am I looking this wrong ? 🤔

Reply:
This isn't about empty or not, but whether `Content-Type` is already set or not.

In a `dict`, `setdefault` only updates `Content-Type` if it does not exist already. This is exactly the behavior we want.

However, `HTTPHeaderDict` does not have a `setdefault` method. Maybe we can add it, but if we don't we can't call `setdefault` here, we need to do this manually:

```
if "Content-Type" not in headers:
    headers ["Content-Type"] = "application/json"
```

And that's it! The code will work for both dict and HTTPHeaderDict.

Does that make sense?

Reply:
Yes. It makes sense. Thank you so much